src/field/vecops.cxx

```diff
@@ -82,35 +81,33 @@ const Vector3D Grad(const Field3D &f, CELL_LOC outloc) {
     return Grad(f, CELL_XLOW, CELL_YLOW, CELL_ZLOW);
   }
 
-  if (outloc == CELL_DEFAULT) {
-    outloc = f.getLocation();
-  }
-
   return Grad(f, outloc, outloc, outloc);
 }
 
-const Vector3D Grad_perp(const Field3D &f, CELL_LOC outloc_x, CELL_LOC UNUSED(outloc_y),
+const Vector3D Grad_perp(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
                          CELL_LOC outloc_z) {
-  Vector3D result(f.getMesh());
-
   TRACE("Grad_perp( Field3D )");
 
-  if (outloc_x == CELL_DEFAULT) {
-    outloc_x = f.getLocation();
-  }
-  if (outloc_z == CELL_DEFAULT) {
-    outloc_z = f.getLocation();
-  }
+  Vector3D result(f.getMesh());
 
-  Coordinates *metric_x = mesh->coordinates(outloc_x);
-  Coordinates *metric_z = mesh->coordinates(outloc_z);
+  ASSERT1((outloc_x == outloc_y && outloc_x == outloc_z) ||
+          (outloc_x == CELL_XLOW && outloc_y == CELL_YLOW &&
+           outloc_z == CELL_ZLOW)); // CELL_VSHIFT
+
+  Coordinates *metric_x = f.getCoordinates(outloc_x);
+  Coordinates *metric_z = f.getCoordinates(outloc_z);
 
   result.x = DDX(f, outloc_x) -
              metric_x->g_12 * DDY(f, outloc_x) / SQ(metric_x->J * metric_x->Bxy);
   result.y = 0.0;
   result.z = DDZ(f, outloc_z) -
              metric_z->g_23 * DDY(f, outloc_z) / SQ(metric_z->J * metric_z->Bxy);
 
+  CELL_LOC outloc = (outloc_x == outloc_y && outloc_x == outloc_z)
+                        ? result.x.getLocation()
+                        : CELL_VSHIFT;
+  result.setLocation(outloc);
```

Comment:
Do we also need to make sure the individual components are now all at the same/consistent location as well? (or `result.y` at least?)

Reply:
Nope, `result.setLocation` sets the location on each of the components.